rust/lance-index/src/util/datafusion.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use arrow_schema::DataType;
+use datafusion_common::ScalarValue;
+
+// This is slightly tedious but when we convert expressions from SQL strings to logical
+// datafusion expressions there is no type coercion that happens.  In other words "x = 7"
+// will always yield "x = 7_u64" regardless of the type of the column "x".  As a result, we
+// need to do that literal coercion ourselves.
+pub(crate) fn safe_coerce_scalar(value: &ScalarValue, ty: &DataType) -> Option<ScalarValue> {
```

Comment:
Is this the same as: https://github.com/lancedb/lance/blob/6fb62bd5d57f569049303118ea9002f70e863c8d/rust/lance/src/datafusion/logical_expr.rs#L27

Should we merge with that?

Reply:
Good find, yes.  I will merge these.

Reply:
There's also: https://docs.rs/datafusion/latest/datafusion/optimizer/simplify_expressions/expr_simplifier/struct.ExprSimplifier.html#method.coerce

which I think does a similar thing. I'll be putting up a PR soon that adds it into our planning.

Reply:
For now, I've updated `resolve_value` to call into `safe_coerce_scalar`.  I updated `safe_coerce_scalar` so it could convert int -> float to match the behavior of `resolve_value`.